File: svm/svm.R
# Importing the dataset
dataset = read.csv('Iris_Data.csv')

# Encoding categorical data
dataset$Species = as.factor(dataset$Species)

# Feature Scaling
scaled_cols = scale(dataset[, 1:2])
dataset[, 1:2] = scaled_cols

# Fitting SVM model
library(e1071)
classifier = svm(formula = Species ~ .,
                 data = dataset,
                 type = 'C-classification',
                 kernel = 'radial')
                 
# Prediction
y_pred = predict(classifier, newdata = dataset)

# Showing the Confusion Matrix and Accuracy
library(caret)
cm = confusionMatrix(y_pred, dataset$Species)
print(cm$table)
print(cm$overall['Accuracy'])

# Visualizing the results
set = dataset
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Sepal.Length', 'Sepal.Width')
y_grid = predict(classifier, newdata = grid_set)
plot(NULL,
     main = 'SVM',
     xlab = 'Sepal Length (Scaled)', ylab = 'Sepal Width (Scaled)',
     xlim = range(X1), ylim = range(X2))
points(grid_set, pch = 20, col = c("#e63946", "#f1fa8c", "#06d6a0")[y_grid])
points(set, pch = 21, bg = c("#e6394699", "#f1fa8c99", "#06d6a099")[dataset$Species])

File: svm.py
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Iris_Data.csv')
X = dataset.iloc[:, :-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)

# one hot encoder is for independent, dependent is labelencoder
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y)

# Training the Kernel SVM model
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf')
classifier.fit(X, y)

y_pred = classifier.predict(X)

# Showing the Confusion Matrix and Accuracy
from sklearn.metrics import confusion_matrix, accuracy_score
print(confusion_matrix(y, y_pred))
print(accuracy_score(y, y_pred))


# Visualizing the results
from matplotlib.colors import ListedColormap
X_set, y_set = X, y
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1,
                               stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1,
                               stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, classifier.
             predict(np.array([X1.flatten(), X2.flatten()]).T).
             reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(['#e63946', '#f1fa8c', '#06d6a0']))
for i in np.unique(y_set):
    plt.scatter(X_set[y_set == i, 0], X_set[y_set == i, 1],
                color = ListedColormap(['#e63946', '#f1fa8c', '#06d6a0'])(i),
                edgecolors = 'black')
plt.title('SVM')
plt.xlabel('Sepal Length (Scaled)')
plt.ylabel('Sepal Width (Scaled)')
plt.show()